### PR TITLE
Paste from word should work for legal list only

### PR DESCRIPTION
**Pull request by illia-stv**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Fix (Paste-from-office): When paste from word should recognise legal list only. Closes [#6019](https://github.com/cksource/ckeditor5-commercial/issues/6019).

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

**Review by mmotyczynska — CHANGES_REQUESTED** [review]
Works as expected.

I tried to find all variables containing `legal` word and propose a change so they are more descriptive. If you find more please also change them so they contain `legalStyleList` phrase rather than simply `legal` or `legalType` (as now we treat it as a style, not a type).

Other than that, please also add a dedicated test with a Word file containing an example of legal-style list and multi-level lists.

**Comment by illia-stv** [comment]
> Works as expected.
> 
> I tried to find all variables containing `legal` word and propose a change so they are more descriptive. If you find more please also change them so they contain `legalStyleList` phrase rather than simply `legal` or `legalType` (as now we treat it as a style, not a type).
> 
> Other than that, please also add a dedicated test with a Word file containing an example of legal-style list and multi-level lists.

There are already Word files which contain all kinds of multi list (legal lists included). In case with legal list our logic adds `legal-list` class, and current test modification cover that case as other cases with multi level (not legal) list which should have behaviour as previously.

**Comment by Witoso** [comment]
Should we leak this class if someone has not enabled Multi-level list plugin? I feel this may be unintended in a case when a plugin is not used.

File: packages/ckeditor5-paste-from-office/src/filters/list.ts [hunk 1]
```diff
@@ -31,7 +31,8 @@ import {
  */
 export function transformListItemLikeElementsIntoLists(
 	documentFragment: ViewDocumentFragment,
-	stylesString: string
+	stylesString: string,
+	isMultiLevelListPlugin: boolean
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/filters/list.ts:35]:
```suggestion
    hasMultiLevelListPlugin: boolean
```

File: packages/ckeditor5-paste-from-office/src/filters/list.ts [hunk 2]
```diff
@@ -214,16 +218,16 @@ function detectListStyle( listLikeItem: ListLikeElement, stylesString: string )
 			}
 		}
 
-		if ( legalListMatch ) {
-			// type = 'ol'; // TODO: while working on the task: PFW should work for legal only
+		if ( islegalStyleList ) {
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/filters/list.ts:221]:
```suggestion
		if ( isLegalStyleList ) {
```

File: packages/ckeditor5-paste-from-office/src/filters/list.ts [hunk 3]
```diff
@@ -214,16 +218,16 @@ function detectListStyle( listLikeItem: ListLikeElement, stylesString: string )
...
 		}
 	}
 
 	return {
 		type,
 		startIndex,
-		style: mapListStyleDefinition( listStyleType )
-		// isLegal: !!legalListMatch // TODO: while working on the task: PFW should work for legal only
+		style: mapListStyleDefinition( listStyleType ),
+		isLegalStyleList: islegalStyleList
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/filters/list.ts:230]:
```suggestion
		isLegalStyleList
```

File: packages/ckeditor5-paste-from-office/src/filters/list.ts [hunk 4]
```diff
@@ -61,12 +62,12 @@ export function transformListItemLikeElementsIntoLists(
 			const listStyle = detectListStyle( itemLikeElement, stylesString );
 
 			if ( !currentList ) {
-				currentList = insertNewEmptyList( listStyle, itemLikeElement.element, writer );
+				currentList = insertNewEmptyList( listStyle, itemLikeElement.element, writer, isMultiLevelListPlugin );
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/filters/list.ts:65]:
```suggestion
				currentList = insertNewEmptyList( listStyle, itemLikeElement.element, writer, hasMultiLevelListPlugin );
```

File: packages/ckeditor5-paste-from-office/src/filters/list.ts [hunk 5]
```diff
@@ -61,12 +62,12 @@ export function transformListItemLikeElementsIntoLists(
...
 			} else if ( itemLikeElement.indent > currentIndentation ) {
 				const lastListItem = currentList.getChild( currentList.childCount - 1 ) as ViewElement;
 				const lastListItemChild = lastListItem!.getChild( lastListItem.childCount - 1 ) as ViewElement;
 
-				currentList = insertNewEmptyList( listStyle, lastListItemChild, writer );
+				currentList = insertNewEmptyList( listStyle, lastListItemChild, writer, isMultiLevelListPlugin );
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/filters/list.ts:70]:
```suggestion
				currentList = insertNewEmptyList( listStyle, lastListItemChild, writer, hasMultiLevelListPlugin );
```

File: packages/ckeditor5-paste-from-office/src/filters/list.ts [hunk 6]
```diff
@@ -178,9 +179,12 @@ function detectListStyle( listLikeItem: ListLikeElement, stylesString: string )
...
 	const listStyleTypeRegex = /mso-level-number-format:([^;]{0,100});/gi;
 	const listStartIndexRegex = /mso-level-start-at:\s{0,100}([0-9]{0,10})\s{0,100};/gi;
-	const legalListTypeRegex = new RegExp( `@list l${ listLikeItem.id }:level\\d\\s*{[^{]*mso-level-text:"%\\d\\\\.`, 'gi' );
+	const legalStyleListRegex = new RegExp( `@list l${ listLikeItem.id }:level\\d\\s*{[^{]*mso-level-text:"%\\d\\\\.`, 'gi' );
+	const multiLevelNumberFormatTypeRegex = new RegExp( `@list l${ listLikeItem.id }:level\\d\\s*{[^{]*mso-level-number-format:`, 'gi' );
 
-	const legalListMatch = legalListTypeRegex.exec( stylesString );
+	const legalStyleListMatch = legalStyleListRegex.exec( stylesString );
+	const multiLevelNumberFormatMatch = multiLevelNumberFormatTypeRegex.exec( stylesString );
+	const islegalStyleList = legalStyleListMatch && !multiLevelNumberFormatMatch;
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/filters/list.ts:187]: Maybe we could also avoid checking this if the Multi-level list plugin is not loaded?

File: packages/ckeditor5-paste-from-office/src/filters/list.ts [hunk 7]
```diff
@@ -324,7 +328,8 @@ function mapListStyleDefinition( value: string ) {
 function insertNewEmptyList(
 	listStyle: ReturnType<typeof detectListStyle>,
 	element: ViewElement,
-	writer: UpcastWriter
+	writer: UpcastWriter,
+	isMultiLevelListPlugin: boolean
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/filters/list.ts:332]:
```suggestion
	hasMultiLevelListPlugin: boolean
```

File: packages/ckeditor5-paste-from-office/src/filters/list.ts [hunk 8]
```diff
@@ -342,9 +347,9 @@ function insertNewEmptyList(
 		writer.setAttribute( 'start', listStyle.startIndex, list );
 	}
 
-	// if ( listStyle.isLegal ) { // TODO: while working on the task: PFW should work for legal only
-	// 	writer.addClass( 'legal-list', list );
-	// }
+	if ( listStyle.isLegalStyleList && isMultiLevelListPlugin ) {
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/filters/list.ts:350]:
```suggestion
	if ( listStyle.isLegalStyleList && hasMultiLevelListPlugin ) {
```

File: packages/ckeditor5-paste-from-office/src/normalizers/mswordnormalizer.ts [hunk 9]
```diff
@@ -22,13 +23,16 @@ const msWordMatch2 = /xmlns:o="urn:schemas-microsoft-com/i;
 export default class MSWordNormalizer implements Normalizer {
 	public readonly document: ViewDocument;
 
+	public readonly editor: Editor;
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/normalizers/mswordnormalizer.ts:26]: Do we need `editor` here? Why don't just store only information that we need, i.e. `hasMultiLevelListPlugin`?

File: packages/ckeditor5-paste-from-office/src/normalizers/mswordnormalizer.ts [hunk 10]
```diff
@@ -22,13 +23,16 @@ const msWordMatch2 = /xmlns:o="urn:schemas-microsoft-com/i;
...
+
 	/**
 	 * Creates a new `MSWordNormalizer` instance.
 	 *
 	 * @param document View document.
 	 */
-	constructor( document: ViewDocument ) {
+	constructor( document: ViewDocument, edytor: Editor ) {
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/normalizers/mswordnormalizer.ts:33]: It's `editor`, not `edytor`. This is a second typo I've noticed in this review. Maybe you could be a bit more careful when you write next time? :wink:

File: packages/ckeditor5-paste-from-office/src/normalizers/mswordnormalizer.ts [hunk 11]
```diff
@@ -43,8 +47,9 @@ export default class MSWordNormalizer implements Normalizer {
 	 */
 	public execute( data: NormalizerData ): void {
 		const { body: documentFragment, stylesString } = data._parsedData;
+		const multiLevelListPlugin = this.editor.plugins.has( 'MultiLevelList' );
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/normalizers/mswordnormalizer.ts:50]:
```suggestion
		const hasMultiLevelListPlugin = this.editor.plugins.has( 'MultiLevelList' );
```

File: packages/ckeditor5-paste-from-office/src/normalizers/mswordnormalizer.ts [hunk 12]
```diff
@@ -43,8 +47,9 @@ export default class MSWordNormalizer implements Normalizer {
...
 
-		transformListItemLikeElementsIntoLists( documentFragment, stylesString );
+		transformListItemLikeElementsIntoLists( documentFragment, stylesString, multiLevelListPlugin );
```
Comment (mmotyczynska) [inline packages/ckeditor5-paste-from-office/src/normalizers/mswordnormalizer.ts:52]:
```suggestion
		transformListItemLikeElementsIntoLists( documentFragment, stylesString, hasMultiLevelListPlugin );
```

**Review by mmotyczynska — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-paste-from-office/src/filters/list.ts (hunk 6 above)
Comment (illia-stv) [reply packages/ckeditor5-paste-from-office/src/filters/list.ts:187]: I think it should not cause performance issue, and also if multi level list plugin is not loaded it doesn't change the fact legal list was detected. It just give us information legal list was found and thats it. Finally our logic will check does multi level list plugin is loaded or not, so checking this 2 times to my mind is not so necessary.

Comment (niegowski) [inline packages/ckeditor5-paste-from-office/src/filters/list.ts:187]: Please add a code comment to explain why this works.

File: packages/ckeditor5-paste-from-office/src/normalizers/mswordnormalizer.ts [hunk 13]
```diff
@@ -22,13 +22,16 @@ const msWordMatch2 = /xmlns:o="urn:schemas-microsoft-com/i;
...
 
+	public readonly hasMultiLevelListPlugin: boolean;
+
 	/**
 	 * Creates a new `MSWordNormalizer` instance.
 	 *
 	 * @param document View document.
 	 */
-	constructor( document: ViewDocument ) {
+	constructor( document: ViewDocument, hasMultiLevelListPlugintor: boolean ) {
```
Comment (niegowski) [inline packages/ckeditor5-paste-from-office/src/normalizers/mswordnormalizer.ts:32]:
```suggestion
	constructor( document: ViewDocument, hasMultiLevelListPlugin: boolean ) {
```

File: packages/ckeditor5-paste-from-office/tests/filters/list.js [hunk 14]
```diff
@@ -104,6 +107,38 @@ describe( 'PasteFromOffice - filters', () => {
 				);
 			} );
 
+			it( 'handles "legal-list" when multi-level-list is loaded', () => {
+				const level1 = 'style="mso-list:l0 level1 lfo0"';
+				const styles = '@list l0:level1\n' +
+					'{mso-level-text:"%1\\.%2\\.";}';
```
Comment (niegowski) [inline packages/ckeditor5-paste-from-office/tests/filters/list.js:113]: This is not verifying the regexp. I'd suggest adding more cases or adding a payload-based test.

**Review by niegowski — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-paste-from-office/src/filters/list.ts [hunk 15]
```diff
@@ -179,11 +179,17 @@ function detectListStyle( listLikeItem: ListLikeElement, stylesString: string )
...
 	const multiLevelNumberFormatTypeRegex = new RegExp( `@list l${ listLikeItem.id }:level\\d\\s*{[^{]*mso-level-number-format:`, 'gi' );
 
 	const legalStyleListMatch = legalStyleListRegex.exec( stylesString );
 	const multiLevelNumberFormatMatch = multiLevelNumberFormatTypeRegex.exec( stylesString );
+
+	/**
+	 * Multi level lists in Word have mso-level-number-format attribute except legal lists,
+	 * so we used that. If list has legal list match and doesn't has mso-level-number-format
+	 * then this is legal-list.
+	 */
```
Comment (niegowski) [inline packages/ckeditor5-paste-from-office/src/filters/list.ts:192]:
```suggestion
	// Multi level lists in Word have mso-level-number-format attribute except legal lists,
	// so we used that. If list has legal list match and doesn't has mso-level-number-format
	// then this is legal-list.
```

File: packages/ckeditor5-paste-from-office/src/filters/list.ts [hunk 16]
```diff
@@ -179,11 +179,17 @@ function detectListStyle( listLikeItem: ListLikeElement, stylesString: string )
 	const listStyleRegexp = new RegExp( `@list l${ listLikeItem.id }:level${ listLikeItem.indent }\\s*({[^}]*)`, 'gi' );
 	const listStyleTypeRegex = /mso-level-number-format:([^;]{0,100});/gi;
 	const listStartIndexRegex = /mso-level-start-at:\s{0,100}([0-9]{0,10})\s{0,100};/gi;
-	const legalStyleListRegex = new RegExp( `@list l${ listLikeItem.id }:level\\d\\s*{[^{]*mso-level-text:"%\\d\\\\.`, 'gi' );
+	const legalStyleListRegex = new RegExp( `@list\\s{1,}l${ listLikeItem.id }:level\\d\\s*{[^{]*mso-level-text:"%\\d\\\\.`, 'gi' );
```
Comment (niegowski) [inline packages/ckeditor5-paste-from-office/src/filters/list.ts:182]: Why such a change? Does some real payload fail on this? If so then what about the above regexp? Should we update it too?

Comment (niegowski) [reply packages/ckeditor5-paste-from-office/src/filters/list.ts:182]:
```suggestion
	const legalStyleListRegex = new RegExp( `@list\\s+l${ listLikeItem.id }:level\\d\\s*{[^{]*mso-level-text:"%\\d\\\\.`, 'gi' );
```

**Review by niegowski — APPROVED** [review]

**Pull request merged**